Sources/GRPCCore/Streaming/Internal/BroadcastAsyncSequence.swift

```diff
@@ -0,0 +1,1734 @@
+/*
+ * Copyright 2023, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import DequeModule
+
+/// An `AsyncSequence` which can broadcast its values to multiple consumers concurrently.
+///
+/// The sequence is not a general-purpose broadcast sequence; it is tailored specifically for the
+/// requirements of gRPC Swift, in particular it is used to support retrying and hedging requests.
+///
+/// In order to achieve this it maintains on an internal buffer of elements which is limited in
+/// size. Each iterator ("subscriber") maintains an offset into the elements which the sequence has
+/// produced over time. If a subscriber is consuming too slowly (and the buffer is full) then the
+/// sequence will cancel the subscriber's subscription to the stream, dropping the oldest element
+/// in the buffer to make space for more elements. If the buffer is full and all subscribers are
+/// equally slow then all producers are suspended until the buffer drops to a reasonable size.
+///
+/// The expectation is that the number of subscribers will be low; for retries there will be at most
+/// one subscriber at a time, for hedging there may be at most five subscribers at a time.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+@usableFromInline
+struct BroadcastAsyncSequence<Element: Sendable>: Sendable, AsyncSequence {
```

Comment:
Overall I am a bit surprised that this is a root asynchronous sequence and not a transformational one. In my mind `broadcast` should always be transformational async sequence that you can apply to something like an `AsyncStream` or `AsyncBufferredStream` is there any particular reason why this is a root asynchronous sequence?

Reply:
It better fits the use case here. A general purpose one probably would be transformational rather than a root. Mostly I want to avoid double buffering, but also since requests have a provider rather than an async sequence we can avoid a bunch of allocations by just creating the one async sequence.